lisa/sut_orchestrator/libvirt/platform.py

```diff
@@ -1341,3 +1388,316 @@ def _capture_libvirt_logs(self) -> None:
+                pool_devices.append(device)
+                pool[iommu_grp] = pool_devices
+            self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+
+    def _verify_device_passthrough_post_boot(
+        self,
+        node_context: NodeContext,
+    ) -> None:
+        device_context = node_context.device_passthrough_context
+        for context in device_context:
+            devices = context.device_list
+            for device in devices:
+                err = f"Kernel driver is not vfio-pci for device: {device}"
+                pool_type = context.pool_type.value
+                if context.managed == "yes" and "pci" in pool_type:
+                    is_vfio_pci = self._is_driver_vfio_pci(device)
+                    assert is_vfio_pci, err
+
+    def _check_passthrough_support(self, node: Node) -> None:
+        ls = node.tools[Ls]
+        path = "/dev/vfio/vfio"
+        err = "Host does not support IOMMU"
+        if not ls.path_exists(path=path, sudo=True):
+            raise LisaException(f"{err} : {path} does not exist")
+
+        path = "/sys/kernel/iommu_groups/"
+        if len(ls.list(path=path, sudo=True)) == 0:
+            raise LisaException(f"{err} : {path} does not have any entry")
+
+    def _is_driver_vfio_pci(
+        self,
+        device_addr: DeviceAddressSchema,
+    ) -> bool:
+        lspci = self.host_node.tools[Lspci]
+        device_addr_str = self._get_pci_address_str(device_addr)
+        kernel_module = lspci.get_used_module(device_addr_str)
+        return kernel_module == "vfio-pci"
+
+    def _set_device_passthrough_node_context(
+        self,
+        node_context: NodeContext,
+        node_runbook: BaseLibvirtNodeSchema,
+    ) -> None:
+        if node_runbook.device_passthrough:
+            node_context.is_device_passthrough_set = True
+            for config in node_runbook.device_passthrough:
+                device_context = DevicePassthroughContext()
+                device_context.managed = config.managed
+                device_context.pool_type = config.pool_type
+                devices = self._get_devices_from_pool(config.pool_type, config.count)
+                device_context.device_list = devices
+                node_context.device_passthrough_context.append(device_context)
+
+    def _create_device_pool(
+        self,
+        pool_type: HostDevicePoolType,
+        vendor_id: str,
+        device_id: str,
+    ) -> None:
+        self.AVAILABLE_HOST_DEVICES[pool_type] = {}
+        lspci = self.host_node.tools[Lspci]
+        device_list = lspci.get_devices_by_vendor_device_id(
+            vendor_id=vendor_id,
+            device_id=device_id,
+        )
+        primary_nic_iommu = self._get_primary_nic_id()
+        for item in device_list:
+            device = self.POOL_TYPE_TO_DEVICE_PROPERTY[pool_type]()
+            slot_info = item.slot.split(":")
+
+            device.domain = item.domain
+            device.bus = slot_info[0]
```

Comment:
You should have a `_parse_pci_address_str` function and place it next to the `_get_pci_address_str` function.

Reply:
@cwize1 
Sure, I will create it as you suggested.

There is only one occurrence to parse the string of PCI address to get bus, slot, function. That is why I did not create the separate function for it.

Reply:
Even if something is only used once, sometimes moving it into a function can help with readability.

Also, in this case, there is an opposite function. So, moving into a function can help keep related code together.

Reply:
@cwize1 Implemented `_parse_pci_address_str `. Please check